src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -640,7 +640,7 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 	}
 
 	/* extract range table entries */
-	ExtractRangeTableEntryWalker((Node *) queryTree, &rangeTableList);
+	ExtractRangeTableEntryWalker((Node *) originalQuery, &rangeTableList);
```

Comment:
I think this change makes sense in general. But, by doing that, we're now treating pulled up subqueries a little different.

Before the patch:
```SQL

DELETE FROM stores_purchase WHERE  store_id > (SELECT logicalrelid  FROM pg_dist_partition);
ERROR:  relation pg_dist_partition is not distributed
```
After the patch:
```SQL
 DELETE FROM stores_purchase WHERE  store_id > (SELECT logicalrelid  FROM pg_dist_partition);
DEBUG:  generating subplan 3_1 for subquery SELECT logicalrelid FROM pg_dist_partition
DEBUG:  Plan 3 query after replacing subqueries and CTEs: DELETE FROM public.stores_purchase WHERE ((store_id)::oid OPERATOR(pg_catalog.>) ((SELECT intermediate_result.logicalrelid FROM read_intermediate_result('3_1'::text, 'binary'::citus_copy_format) intermediate_result(logicalrelid regclass)))::oid)
DELETE 0
```

Do you see any case where we'd regress? I think no, but wanted to ask in case you can think of any edges?

Reply:
That error happened in a pretty awkward place because we allowed the local table to by-pass our checks:
```
#0  errstart (elevel=elevel@entry=20, filename=filename@entry=0x7f84f1887f88 "metadata/metadata_cache.c", lineno=lineno@entry=782, 
    funcname=funcname@entry=0x7f84f1889150 <__func__.29196> "DistributedTableCacheEntry", domain=domain@entry=0x0) at elog.c:251
#1  0x00007f84f183680b in DistributedTableCacheEntry (distributedRelationId=distributedRelationId@entry=16799) at metadata/metadata_cache.c:782
#2  0x00007f84f1852a55 in TargetShardIntervalsForRestrictInfo (restrictionContext=<optimized out>, multiShardQuery=multiShardQuery@entry=0x7ffeb3c7a96f, 
    partitionValueConst=partitionValueConst@entry=0x7ffeb3c7a9d0) at planner/multi_router_planner.c:2294
#3  0x00007f84f18533a8 in PlanRouterQuery (originalQuery=originalQuery@entry=0xeeb968, plannerRestrictionContext=plannerRestrictionContext@entry=0xf60568, 
    placementList=placementList@entry=0x7ffeb3c7a9f0, anchorShardId=anchorShardId@entry=0x7ffeb3c7a9f8, relationShardList=relationShardList@entry=0x7ffeb3c7a9e8, 
    prunedShardIntervalListList=prunedShardIntervalListList@entry=0x7ffeb3c7a9e0, replacePrunedQueryWithDummy=true, multiShardModifyQuery=0x7ffeb3c7a9df, 
    partitionValueConst=0x7ffeb3c7a9d0) at planner/multi_router_planner.c:2039
#4  0x00007f84f1853754 in RouterJob (originalQuery=originalQuery@entry=0xeeb968, plannerRestrictionContext=plannerRestrictionContext@entry=0xf60568, 
    planningError=planningError@entry=0x1054b20) at planner/multi_router_planner.c:1649
#5  0x00007f84f1853bff in CreateModifyPlan (originalQuery=originalQuery@entry=0xeeb968, query=query@entry=0xeeba80, 
```

In that sense this is another bug that is fixed. 

I don't think we would regress since the rewritten query contains a subset of RTEs of the original query (since views are already expanded in the original query).